---
title: "GSEA continued + File formats"
output:
  html_document:
    df_print: paged
date: "due April 20, 2021"
---

Name:

## My Name

```{r setup, include=FALSE}
library(tidyverse)
Sys.getenv("HOME") # see where your home directory is
course_directory <- file.path(Sys.getenv("HOME"), "repos", "hugego")
course_directory # double check what you set your course_directory to
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = course_directory)
```

# Objectives

- gain a deeper understanding of GSEA through simulation

- review concepts from lecture on File formats


## Simulated GSEA

In the following code block, we are simulating a ranked genelist with known enrichment in the top part of the list. In our simulation, we want to look at how parameters such as  enrichment cutoff and magnitude affect the resulting GSEA graph.

Take some time to review the code below:

```{r basemodel, eval=FALSE}

# define parameters:
ngenes <- 2e4 # total number of genes
cutoff <- round(ngenes / 10, 0) # number of top ranked genes with enrichment (default here 1/5 of genes)
pathway_genes <- 50 # number of genes in pathway
baseline <- pathway_genes / ngenes # the baseline is the fraction of total genes in pathway of interest, i.e. the expect frequency with no enrichment
enrich <- baseline * 2 # probability of pathway genes in top list

# simulate a dataset:
if (enrich > 1) {warning("total enrichment probability is greater than 100%")}
if (baseline * cutoff < 1) {print("WARNING: baseline probability is too low")}

# First, we create the top ranked genes in our list, which are enriched by the factor `enrich`
top <- sample(c(rep(0, round((1-enrich) * cutoff, 0)), rep(1, round(enrich * cutoff))), 
              size = cutoff, 
              replace = F)

# Then, we will generate the rest of the list, which has the remaining genes in our pathway of interest
if (sum(top) <= baseline * ngenes) {
  bot <- sample(c(rep(0, round((1-baseline) * (ngenes - cutoff), 0) + sum(top)), 
                  rep(1, round(ngenes * baseline, 0) - sum(top))),
                size = ngenes-cutoff, replace = F)
  ranked_list <- c(top, bot)

  ## create a vector to store enrichment scores:
  escore <- numeric(ngenes)

  ## iterate enrichment scores over the ranked list of genes:      ###############
  running_sum <- 0                                                 # Answer      #
  for (i in 1:ngenes) {                                            # question 53 #
    ## the expected number of genes changes with each iteration    # from this   #
    expected = baseline * i                                        # block of    #
    ## ENRICHMENT FUNCTION:                                        # code!       #
    running_sum <- running_sum + ranked_list[i]                    #             #
    escore[i] <- running_sum - expected                            #             #
  }                                                                ###############
  
  # visualize the results:
  plot(x=1:ngenes, y=escore, type='l', ylim=c(-max(abs(escore)), max(abs(escore))))
  abline(h=0)
  points(x = which(as.logical(ranked_list)), y = rep(-max(abs(escore)), sum(ranked_list)))
} else {print("COULD NOT EVAL: NUM HITS IN TOP LIST GREATER THAN PATHWAY")}
```

## Exercise 1

### Question 1.1

In your own words, describe how the enrichment function function works. 

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

With the current settings, what fold-enrichment is there for how many genes?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

### Question 1.2

- Execute the code several more times by clicking the green arrow in the upper right of the code block in rstudio. Would you characterize the resulting graph as stable?

- Try varying the cutoff parameter to increase the fraction of the list with probable enrichment. At what number does it become stable (_i.e._ consistently shows enrichment)? What effect does list length (the cutoff parameter) have on your ability to detect enrichment?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

### Question 1.3

-Try varying the enrichment parameter with a moderate sized list (cutoff=100; _i.e._ set the denominator to 200). At what fold-enrichment (enrich $\div$ baseline) does the graph appear to be stably enriched? Is the sensitivity to fold-enrichment affected by list length?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

## Exercise 2

Suppose we wanted to add a pvalue calculation using Fisher's exact test, using the code block below. <span style="color:red"><i><b>WARNING: THIS IS NOT HOW WE CALCULATE SIGNIFICANCE IN GSEA. IT IS FOR EXPLORATION PURPOSES ONLY. WE WILL TRY A SIGNIFICANCE CALCULATION IN EXERCISE 3.</b></i></span>


```{r pvalues, eval=FALSE}

# define parameters:
ngenes <- 2e4 # total number of genes
cutoff <- round(ngenes / 5, 0) # number of top ranked genes with enrichment (default here 1/5 of genes)
pathway_genes <- 50 # number of genes in pathway
baseline <- pathway_genes/ngenes # fraction of total genes in pathway of interest
enrich <- baseline * 2 # probability of pathway genes in top list

# simulate a dataset:
if (enrich > 1) {warning("total enrichment probability is greater than 100%")}
if (baseline * cutoff < 1) {print("WARNING: baseline probability is too low")}

top <- sample(c(rep(0, round((1 - enrich) * cutoff, 0)), 
                rep(1, round(enrich * cutoff, 0))), 
              size = cutoff, 
              replace = F)

if (sum(top) <= baseline * ngenes) {
  bot <- sample(c(rep(0, round((1-baseline) * (ngenes - cutoff), 0) + sum(top)), 
                  rep(1, round(ngenes * baseline, 0) - sum(top))),
                size = ngenes-cutoff, 
                replace = F)
  ranked_list <- c(top, bot)

  ## create vectors to store enrichment scores and pvals:
  escore <- numeric(ngenes)
  epval  <- numeric(ngenes)
  
  ## iterate enrichment scores over the ranked list of genes:
  running_sum <- 0
  for (i in 1:ngenes) {
    expected = round(baseline * i, 0)
    ## ENRICHMENT FUNCTION:
    running_sum <- running_sum + ranked_list[i]
    escore[i] <- running_sum - expected
    ### uncomment this section for exercise 2.2
    # epval[i] <- phyper(
    #   ## REMINDER! the upper tail is non-inclusive
    #   q = # enter code here & uncomment,
    #   m = # enter code here & uncomment,
    #   n = # enter code here & uncomment,
    #   k = # enter code here & uncomment,
    #   lower.tail = # enter code here & uncomment,
    #   log.p = # enter code here & uncomment
    # )
  }
  
  # visualize the results:
  plot(x=1:ngenes, y=escore, type='l', ylim=c(-max(abs(escore)), max(abs(escore))))
  abline(h=0)
  points(x = which(as.logical(ranked_list)), y = rep(-max(abs(escore)), sum(ranked_list)))
} else {print("COULD NOT EVAL: NUM HITS IN TOP LIST GREATER THAN PATHWAY")}
```

### 2.1

Alter the code block above with some parameters that appear to show clear enrichment from exercise 1.3.

### 2.2

Fix the part of the code pertaining to the pvalue calculation by adding arguments to the phyper function. Take a minute to review the code block below. After running the code below, are any of the calculated p values "significant"? 

```{r significance1, eval=FALSE}
print(cutoff)
print(min(which(escore == max(escore))))
print(epval[min(which(escore == max(escore)))])
```

### 2.3 

After running the code above, are any of the calculated p values "significant"? 

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

### 2.4 

What do the three values tell you? Are you surprised by the second result? What conclusions can you draw? Is there a strong relationship between the cutoff and the highest point of enrichment?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

## Exercise 3 Significance Calculation

Let's define a function to simulate a randomized dataset with the same general properties as defined above.

```{r random-gsea-function}

## calculates the maximum enrichment score of a random datasdet
randscore <- function(ngenes, baseline) {
  genes <- c(rep(1, round(baseline * ngenes, 0)), rep(0, ngenes-round(baseline*ngenes, 0)))
  ranked_random_geneset <- sample(genes, size = ngenes, replace = F)
  escore <- numeric(ngenes)
  running_sum <- 0
  tally <- 0
  for (i in 1:ngenes) {
    expected = round(baseline * i, 0)
    running_sum <- running_sum + ranked_random_geneset[i]
    ## ENRICHMENT FUNCTION:
    escore[i] <- running_sum - expected
  }
  max(escore)
}

```

We can use this function to permute the range of scores we can expect with a pathway of size `ngenes * baseline`:

```{r permute-distrib, eval=F}
permutations <- 1000
trialdata <- numeric(permutations)
  
print("This may take a while!.......")
for (i in 1:permutations) {
  trialdata[i] <- randscore(ngenes = ngenes, baseline = baseline)
  if (i %% 100 == 0) print(i)
}

sigcount <- numeric(permutations)
for (i in 1:permutations) {sum(trialdata[i] > escore)}
sum(sigcount) / permutations

trialdata <- data.frame(
  trialnum = 1:permutations,
  maxscore = trialdata)

ggplot(trialdata, aes(x=maxscore)) + geom_histogram()

```

### 3.1

How does your `max(escore)` (from the graph in exercise 2) compare to this histogram?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

### 3.2

Calculate a P-value. To do this, simply quantify the number of random scores ≥ to your `max(escore)` value and divide by the number of permutations:

```{r calc-gsea-pval}
### your code here
```

A major caveat is that this p-value only has as much resolution as 1 ÷ the number of permutations. Therefore if your answer above is equivalent to `0`, how would you state the p value for accuracy? (_i.e._ $p < ?$)

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

### 4

Recall that a FASTQ file contains 4 lines per entry.

For the following entries, what position in the first read is most confidently called? what is it's Quality score (Q) and Probability of incorrect call (P)?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

For the following entries, what position in the second read is least confidently called? What is it's Quality score (Q) and Probability of incorrect call (P)?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>

```
@SRR794330.2 HWI-ST434:134117522:C1N85ACXX:8:1101:1493:2158/1
TGGCTTTGAAGAAGGAGGATGGGGCCACCAGCCAAGGAATGCAGGGAGCCTCTAGAAATTAGAAAAGGCAAGGCAACAGATTCTCCCCTAAAGCCTCCAG
+
;=?ADDD?<4C4;CEFEGE@@7@+@C;EDB)CDDE9B@>?D9??@FA<@@FAE@=D@EEE.)=?BB75=@@1;2(5;<9?>@@B>A:@B?@@########
@SRR794330.3 HWI-ST434:134117522:C1N85ACXX:8:1101:1684:2048/1
ATACAAAAATTAGCTGGGCATGGTGGTGTGCACCTGTAATCCCAGCTACTTGGGAAGCTGAGGCAGGAGAATCGCTTGAACCTGGGAGGTAGAGGTTGCA
+
<@<BBADABHHFFIJIIJG>FHGHGHCFFGIIJJJIAGADFGGIFHD@DDGHEICGA@FFAHGECC>CD?@;>AC@A??AABC2???@@2>@CC?:?CB@
```

### 5

Let's say you have aligned some RNA-Seq, and now you have a BAM file. You would like to know how many reads you have for each gene in your sample's genome.

What is one tool that can be used to count gene abundance, and what type of file would you need to describe the the position and structure of those genes.

<span style="color:#31a354"><i><b>
answer here
</b></i></span>


### 6

Post Alignment QC question (sources will be given after Thursday (04/15/21) lecture - 

Using the provided pre- and post- alignment metrics, evaluate the sequencing files corresponding to your transcription factor. For each metric, see if you can find a sample that looks as good or better, and another that is worse. Describe how you evaluated "better" or "worse". Given only the quality of the data for your transcription factor, would you request a repeat of the experiment if it was in your power to do so?

<span style="color:#31a354"><i><b>
answer here
</b></i></span>
