---
title: 'Lab Assignment I: Install GENAVi'
output:
  pdf_document: default
  html_document:
    df_print: paged
date: "due April 25, 2021"
---

```{r setup, include=FALSE}
Sys.getenv("HOME") # see where your home directory is
course_directory <- file.path(Sys.getenv("HOME"), "repos", "hugego")
course_directory # double check what you set your course_directory to
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(root.dir = course_directory)
```

## Doing this lab assignment

You can do this assignment in one of two ways. 1. You may type commands from the html version of this document into an R console and refer to the exercises or 2. you can open the markdown in Rstudio and enter commands interactively. With this latter option you may edit the header and the "setup" code chunk above with your personal information, annotate the rest of the file with notes and reminders, and enter your answers to questions on the .Rmd file for safe keeping of your notes. **compiled Rmarkdown is an acceptable format to hand in homework.**

## Installing R

[Detailed instructions for all platforms](https://cran.R-project.org). This notebook was composed with R version 4.0.4 "Lost Library Book".

## Installing GENAVi

Prerequisites: Any platform (typically mac, windows or linux), R, Rstudio, and bioconductor.

### Install Bioconductor

If you haven't done so already, install bioconductor. GENAVi will attempt to install a bunch of dependencies when you first run the app, so pre-install those packages so we can troubleshoot this step seperately. The following lines will install all the dependencies that you need to run GENAVi. If you have any difficulty, ask a course instructor for assistance.

```{r installation, echo=TRUE, results='hide', eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
     install.packages("BiocManager")
BiocManager::install(version = "3.12")

pkgs <- c('airway', 'affy', 'affyio', 'annotate', 'AnnotationHub', 'apeglm', 'argonR', 'argonDash', 'ashr', 'bbmle', 'bdsmatrix', 'bs4Dash', 'Cairo', 'classInt', 'clusterProfiler', 'coda', 'cowplot', 'debugme', 'dendextend', 'DESeq2', 'devtools', 'diffobj', 'DO.db', 'DOSE', 'DT', 'e1071', 'enrichplot', 'emdbook', 'edgeR', 'etrunct', 'europepmc', 'fastcluster', 'fastmatch', 'fdrtool', 'fgsea', 'fontquiver', 'fontBitStreamVera', 'fontLiberation', 'freetypeharfbuzz', 'genefilter', 'geneplotter', 'GGally', 'ggdendro', 'ggforce', 'ggplotify', 'ggraph', 'ggrepel', 'GOSemSim', 'graphite', 'graphlayouts', 'gridGraphics', 'GSEABase', 'hexbin', 'highcharter', 'igraph', 'iheatmapr', 'IHW', 'interactiveDisplayBase', 'invgamma', 'IRdisplay', 'irlba', 'KEGG.db', 'limma', 'listviewer', 'locfit', 'lpsymphony', 'maps', 'mixsqp', 'msigdbr', 'mvtnorm', 'SQUAREM', 'maptools', 'numDeriv', 'org.Hs.eg.db', 'parsedated', 'pingr', 'plotly', 'plotlyGeoAssets', 'polyclip', 'prettydoc', 'preprocessCore', 'qvalue', 'REBayes', 'repr', 'rgeos', 'shinydashboardPlus', 'Rhdf5lib', 'reactome.db', 'pasilla', 'quantmod', 'RcppArmadillo', 'reliaR', 'reshape', 'ReactomePA', 'RcppNumerical', 'rhdf5', 'rlist', 'Runit', 'rvcheck', 'sf', 'shinyAce', 'shinyBS', 'shinycssloaders', 'shinyjqui', 'shinytest', 'shinythemes',  'shinyWidgets', 'showimage', 'slam', 'tablerDash', 'tidygraph', 'triebeard', 'truncnorm','TTR', 'tweenr', 'tximport', 'tximportData', 'shinyjs', 'shinydashboard', 'units', 'urltools', 'vdiffr', 'V8', 'vsn', 'webdriver', 'webshot', 'xts', 'zoo')

BiocManager::install(pkgs, update = TRUE, ask = FALSE, checkBuilt = TRUE)
```

For further information or instructions on updating and upgrading bioconductor packages see [Bioconductor website](https://bioconductor.org/install)

### Run GENAVi

The GENAVi package runs on Shiny (a web-based interactive tool for R analyses)
and is available as an [installable R package via
GitHub](https://github.com/alpreyes/GENAVi). It is not necessary to know how to
use git for this excercise, but it is strongly encouraged for you to [learn
github on your own](https://www.coursera.org/learn/version-control-with-git) if
you plan to use bioinformatics analysis for your thesis work.

```{r run-genavi, echo=TRUE, results='hide', eval=FALSE}
shiny::runGitHub("alpreyes/GENAVi")
```

### Running the GENAVi Docker image

1. If needed, [Install Docker](https://docs.docker.com/docker-hub/)

2. After starting the Docker daemon, pull the GENAVi image by running at a terminal prompt (this takes several minutes with a good connection):

```{sh pull-docker, echo=TRUE, eval=FALSE}
docker pull cedarscompbio/genavi
```

2. Run the [Docker image for GENAVi](https://hub.docker.com/r/cedarscompbio/genavi) by following the instructions on the image's web page.

```{sh docker-image-start, eval=FALSE}
docker run --rm -p 3838:3838 cedarscompbio/genavi
```

Then visit the container with your favorite web browser at [localhost:3838/GENAVi/](http://localhost:3838/GENAVi).

### Alternative GENAVi Platform

The Center for Bioinformatics also hosts GENAVi via website [junkdnalab.shinyapps.io/GENAVi](https://junkdnalab.shinyapps.io/GENAVi).

## Exercises

### 1. Create a dataset
You are going to create a dataset consisting of your favorite TF knockdown and appropriate control samples, and upload these datasets into GENAVi.

#### Uploading data to GENAVi

Let's suppose you would like to analyze TEAD2 knockdowns in K562.

In the GENAVi window that you should have opened up by now, you'll notice that there is already a table of genes, experiment and expression values. This is just an example dataset containing expression data from ovarian cancer and ovarian cancer precursor cell lines. Ignore it for now.

1. Using the "Data Upload" Radio button in GENAVi, navigate to the location on your HD containing the file "combinedfeatcounts.csv" and select it. GENAVi will upload the file into its cache, but it may take several minutes depending on your computer. Because it is a '.csv' file, GENAVi automatically recognizes the format of the file. In addition, GENAVi recognizes that you have ENSEMBL gene labels and uses these labels to cross reference public databases and display the human-readable gene symbols. Now click outside the popup to make it disappear.

2. Now click on the Differential Expression Analysis tab along the top (3rd tab from left). This will reveal a new set of radio buttons on the left. Click the first of these, called "Metadata upload" with the excel-spreadsheet icon. Navigate to the location on your computer with the file called "crispri_modelmatrix.csv" and select it for upload. This file has been created for you for this class, but understanding what's in it and how it works with GENAVi's analysis packages is critical to be able to carry out a future independant analysis on your own data. Again, you should click outside the pop up to make it disappear.

3. Notice that this model matrix has 4 columns.

+ Column 1 contains the names of the samples from the data upload section
 
+ Columns 2 & 3 contain experimental covariates. In theory, **you may include as many such columns as you need**. GENAVi allows you to select which variables to analyze.

+ Column 4 indicates the [ensembl geneid](https://uswest.ensembl.org/info/genome/stable_ids/index.html) for the transcription factor that was (ostensibly) knocked down by CRISPRi

+ When you're done, **close the GENAVi window** and check that the console window in Rstudio has returned to the normal R prompt (with '>' carat).
 
4. We need to filter our data set _and_ the metadata (model matrix) so that it contains only the data we want to perform our differential expression on. The data contain every knockdown experiment plus control experiments **AND** ENCODE RNA-seq for _unperturbed_ K562 ("RNAseq" in the treatment column). We can ignore these latter samples for now. Our object is to use the metadata from the model matrix to identify the samples we need from the feature counts '.csv' file and filter both files into smaller versions for direct upload into GENAVi. First we're going to try this in R. For this example I chose TEAD2. You should choose any other transcription factor you like and alter the code as needed.

```{r inspect-data, echo=TRUE, eval=FALSE}
## Let's open the metadata file
setwd(course_directory)
metadat <- read.csv(
  file = file.path(course_directory, 'data', 'crispri_modelmatrix.csv'), 
  header = TRUE,
  stringsAsFactors = FALSE)

## Inspect the TEAD2 samples in the metadata file:
knitr::kable(subset(metadat, treatment == 'TEAD2'))
```

Notice that there are two replicates for this experiment in two different batches. 

a. (easy) After studying the metadata file, is this unique to this factor or is it a general strategy throughout the study?

**type your answer here**

b. (easy) Why do you think the Authors chose the strategy that they did?

**type your answer here**

c. (moderate) Is the strategy appropriate? Does it make sense for analyzing the effects of knockdown of a single factor? Why would (or wouldn't) you recommend this strategy?

**type your answer here**

d. (hard) What types of questions could one answer with this data? Are there designs that would effectively address different study goals?

**type your answer here**

5. Filter your data in R using this code chunk as a guide:

```{r filter-data, echo=TRUE, eval=FALSE}

genesymbol <- 'NSD2'
genebatch <- metadat$batch[metadat$treatment==genesymbol][1]

gene.metadat <- rbind(
  subset(
    subset(metadat, treatment == genesymbol),
    batch == genebatch),
  subset(
    subset(metadat, treatment == 'CRISPRi_CONTROL'),
    batch == genebatch)
  )

gene.samples <- gene.metadat$samples

## Now we read in the data from feature counts file
cts <- read.csv(
  file = file.path(course_directory, 'data', 'combinedfeatcounts.csv'),
  header = TRUE,
  stringsAsFactors = FALSE)
## Take a quick peek at the contents
knitr::kable(cts[1:5, 1:4])

## We need to select the columns from our tead.samples, including the Geneid column:

gene.counts <- subset(cts, select = c("Geneid", gene.samples))

## Now we're ready to export these object for upload to GENAVi
write.table(
  gene.counts, 
  file = file.path(course_directory, 'data', paste(genesymbol, 'featcounts.csv', sep = '')),#'teadfeatcounts.csv'),
  row.names = FALSE,
  quote = FALSE,
  sep = ','
)

write.table(
  gene.metadat,
  file = file.path(course_directory, 'data', paste(genesymbol, 'model.csv', sep = '')),#'teadmodel.csv'),
  row.names = FALSE,
  quote = FALSE,
  sep = ','
)
```

6. Confirm that you can upload your filtered dataset _and_ filtered metadata into GENAVi.



### 2. Doing it the old fashioned way

1. Choose a new TF. Repeat the operation we did in lab exercise 1 with your new TF, but this time use EXCEL or any other commercial or open source spreadsheet program of your choice to filter your data sets and create new data sources. Frequently in scientific analysis settings, you will be handed data from a collaborator or clinician in excel format, which you must be comfortable transforming ('munging') into something more useful for analysis.

2. Confirm that you can import your excel-created data and metadata into GENAVi
